src/main/kotlin/org/vitrivr/cottontail/math/knn/metrics/Keypoints2DDistance.kt

```diff
@@ -0,0 +1,56 @@
+package org.vitrivr.cottontail.math.knn.metrics
+
+import org.vitrivr.cottontail.model.values.DoubleValue
+import org.vitrivr.cottontail.model.values.types.RealVectorValue
+import org.vitrivr.cottontail.model.values.types.VectorValue
+
+/**
+ * Calculates the Keypoints 2D distance between two vectors.
+ * This is the sum of Euclidian distances between consecutive pairs of points in a vector
+ *
+ * @version 1.0
+ * @author Ralph Gasser
+ */
+object Keypoints2DDistance : DistanceKernel {
+    override val cost = 1.0f
+
+    private fun procDiffs(diffs: RealVectorValue<*>): DoubleValue {
+        val sqDiffs = diffs.pow(2)
+        sqDiffs.indices.endInclusive - sqDiffs.indices.start + 1
+        var result = DoubleValue.ZERO
+        for (diffIdxs in sqDiffs.indices.chunked(2)) {
+            result += (sqDiffs[diffIdxs[0]] + sqDiffs[diffIdxs[1]]).sqrt()
+        }
+        return result
+    }
+
+    /**
+     * Calculates the Keypoints 2D distance between two [VectorValue]s.
+     *
+     * @param a First [VectorValue]
+     * @param b Second [VectorValue]
+     *
+     * @return Distance between a and b.
+     */
+    override operator fun invoke(a: VectorValue<*>, b: VectorValue<*>): DoubleValue {
+        if (a !is RealVectorValue || b !is RealVectorValue) {
+            throw ArrayConfigurationNotSupportedException("Keypoints2DDistance only supports RealVectorValues");
+        }
+        if (a.logicalSize )
+        throw ArrayConfigurationNotSupportedException("Keypoints2DDistance only supports RealVectorValues");
+        return procDiffs((a - b) as RealVectorValue<*>)
+    }
+
+    /**
+     * Calculates the weighted Hamming distance between two [VectorValue]s.
+     *
+     * @param a First [VectorValue]
+     * @param b Second [VectorValue]
+     * @param weights A list of weights
+     *
+     * @return Distance between a and b.
+     */
+    override operator fun invoke(a: VectorValue<*>, b: VectorValue<*>, weights: VectorValue<*>): DoubleValue {
+        return procDiffs(weights * (a - b))
```

Comment:
Since you do the square and the root operations within `procDiffs`, shouldn't it rather be `procDiffs((a - b)) * weights` ?

Reply:
Weights seems to be a weight for each dimension (it it to allow for feature selection to be done at query time?) and so it couldn't be applied after procDiffs. It could be applied after squaring but then it would have a different scale. Is that what you're suggesting? I don't know the exact purpose of weights, but this line is analogous to its usage elsewhere.